devdoc/sealable_channel.md

```diff
@@ -0,0 +1,258 @@
+    MOCKABLE_FUNCTION(, int, sealable_channel_open, THANDLE(SEALABLE_CHANNEL), sealable_channel);
+    MOCKABLE_FUNCTION(, void, sealable_channel_close, THANDLE(SEALABLE_CHANNEL), sealable_channel);
+    MOCKABLE_FUNCTION(, SEALABLE_CHANNEL_RESULT, sealable_channel_pull, THANDLE(SEALABLE_CHANNEL), sealable_channel, THANDLE(RC_STRING), correlation_id, ON_SEALABLE_CHANNEL_DATA_AVAILABLE_CB, on_data_available_cb, void*, pull_context);
+    MOCKABLE_FUNCTION(, SEALABLE_CHANNEL_RESULT, sealable_channel_push, THANDLE(SEALABLE_CHANNEL), sealable_channel, THANDLE(RC_STRING), correlation_id, THANDLE(RC_PTR), data, ON_SEALABLE_CHANNEL_DATA_CONSUMED_CB, on_data_consumed_cb, void*, push_context);
+    MOCKABLE_FUNCTION(, int, sealable_channel_get_reporting_info, THANDLE(SEALABLE_CHANNEL), sealable_channel, SEALABLE_CHANNEL_REPORTING_INFO*, reporting_info);
+    MOCKABLE_FUNCTION(, int, sealable_channel_seal_channel, THANDLE(SEALABLE_CHANNEL), sealable_channel);
+```
+
+### sealable_channel_create
+
+```c
+THANDLE(SEALABLE_CHANNEL) sealable_channel_create(THANDLE(PTR(LOG_CONTEXT_HANDLE)) log_context, THANDLE(THREADPOOL) threadpool, uint32_t max_channel_item_count, size_t max_channel_item_size);
+```
+
+`sealable_channel_create` creates a new `SEALABLE_CHANNEL` instance.
+
+If `log_context` is `NULL`, `sealable_channel_create` shall fail and return `NULL`.
+
+If `threadpool` is `NULL`, `sealable_channel_create` shall fail and return `NULL`.
+
+If `max_channel_item_count` is 0, `sealable_channel_create` shall fail and return `NULL`.
+
+If `max_channel_item_size` is 0, `sealable_channel_create` shall fail and return `NULL`.
+
+`sealable_channel_create` shall call `sm_create`.
+
+`sealable_channel_create` shall call `channel_create`.
+
+`sealable_channel_create` shall succeed and return the newly created `SEALABLE_CHANNEL`.
+
+If any error occurs, `sealable_channel_create` shall fail and return `NULL`.
+
+### sealable_channel_open
+
+```c
+int sealable_channel_open(THANDLE(SEALABLE_CHANNEL) sealable_channel);
+```
+
+`sealable_channel_open` opens a `sealable_channel`.
+
+If `sealable_channel` is `NULL`, `sealable_channel_open` shall fail and return a non-zero value.
+
+`sealable_channel_open` shall call `sm_open_begin`.
+
+`sealable_channel_open` shall call `channel_open`.
+
+`sealable_channel_open` shall call `sm_open_end`
+
+`sealable_channel_open` shall succeed and return 0.
+
+If any error occurs, `sealable_channel_open` shall fail and return a non-zero value.
+
+### sealable_channel_close
+
+```c
+void sealable_channel_close(THANDLE(SEALABLE_CHANNEL) sealable_channel);
+```
+
+`sealable_channel_close` closes a `sealable_channel`.
+
+If `sealable_channel` is `NULL`, `sealable_channel_close` shall return immediately.
+
+`sealable_channel_close` shall call `sm_close_begin`.
```

Comment:
If the underlying channel has pending operations, those will not get abandoned automatically. 

sm_close_begin blocks until the pending call count reaches 0. 

I think we need to call channel_close from the sm_exec_begin_with_cb callback to poke the underlying channel.